sql: TestShowRangesMultipleStores failed

sql.TestShowRangesMultipleStores [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21341524?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21341524?buildTab=artifacts#/) on master @ [3360aa3e61d064715b7d634e82afc2a458696bf7](https://github.com/cockroachdb/cockroach/commits/3360aa3e61d064715b7d634e82afc2a458696bf7):


```
=== RUN   TestShowRangesMultipleStores
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/d4bcd057978ca6bf47ca7084fb62a8fa/logTestShowRangesMultipleStores2635471356
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:637: DRPC is enabled (via testserver factory defaults)
    show_ranges_test.go:147: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d4bcd057978ca6bf47ca7084fb62a8fa/logTestShowRangesMultipleStores2635471356
--- FAIL: TestShowRangesMultipleStores (69.51s)
=== RUN   TestShowRangesMultipleStores/SHOW_RANGES_FROM_DATABASE_system_WITH_DETAILS
    soon.go:64: condition failed to evaluate within 45s: from sql_runner.go:401: query '
        SELECT DISTINCT
        		(
        		array_position(replica_localities, lease_holder_locality)
        		= array_position(replicas, lease_holder)
        		)
        	FROM [SHOW RANGES FROM DATABASE system WITH DETAILS]': expected:
        true
        got:
        true
        NULL
        
        goroutine dump: /artifacts/tmp/_tmp/d4bcd057978ca6bf47ca7084fb62a8fa/goroutine_dump_139779690.txt
--- FAIL: TestShowRangesMultipleStores/SHOW_RANGES_FROM_DATABASE_system_WITH_DETAILS (47.93s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestShowRangesMultipleStores.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-63764

sql.TestShowRangesMultipleStores [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21344191?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21344191?buildTab=artifacts#/) on master @ [24a5c4531bfa7ab09dfdb171b4368d42d7c8ae49](https://github.com/cockroachdb/cockroach/commits/24a5c4531bfa7ab09dfdb171b4368d42d7c8ae49):


```
=== RUN   TestShowRangesMultipleStores
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/492cc08314ebb30515b6f60e383090f4/logTestShowRangesMultipleStores822840154
    test_log_scope.go:82: use -show-logs to present logs inline
    show_ranges_test.go:147: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/492cc08314ebb30515b6f60e383090f4/logTestShowRangesMultipleStores822840154
--- FAIL: TestShowRangesMultipleStores (75.47s)
=== RUN   TestShowRangesMultipleStores/SHOW_RANGES_FROM_DATABASE_system_WITH_DETAILS
    soon.go:64: condition failed to evaluate within 45s: from sql_runner.go:401: query '
        SELECT DISTINCT
        		(
        		array_position(replica_localities, lease_holder_locality)
        		= array_position(replicas, lease_holder)
        		)
        	FROM [SHOW RANGES FROM DATABASE system WITH DETAILS]': expected:
        true
        got:
        true
        NULL
        
        goroutine dump: /artifacts/tmp/_tmp/492cc08314ebb30515b6f60e383090f4/goroutine_dump_3596451296.txt
--- FAIL: TestShowRangesMultipleStores/SHOW_RANGES_FROM_DATABASE_system_WITH_DETAILS (48.70s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestShowRangesMultipleStores.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/show_ranges_test.go
+++ b/pkg/sql/show_ranges_test.go
@@ -133,7 +133,12 @@ func TestShowRangesMultipleStores(t *testing.T) {
 		"SHOW RANGE FROM INDEX system.jobs@jobs_status_created_idx FOR ROW ('running', now(), 0)",
 	} {
 		t.Run(q, func(t *testing.T) {
-			// Retry because if there's not a leaseholder, you can get a NULL.
+			// A leaseholder lookup that fails transiently (e.g. while a lease is
+			// being transferred after the SCATTER above) reports lease_holder=0
+			// with a NULL lease_holder_locality, which makes the equality below
+			// evaluate to NULL. Filter those rows out so the assertion only
+			// considers ranges where the leaseholder is known; CheckQueryResultsRetry
+			// will retry until at least one such range exists.
 			sqlDB.CheckQueryResultsRetry(t,
 				fmt.Sprintf(`
 SELECT DISTINCT
@@ -141,7 +146,8 @@ SELECT DISTINCT
 		array_position(replica_localities, lease_holder_locality)
 		= array_position(replicas, lease_holder)
 		)
-	FROM [%s]`, q), [][]string{{"true"}})
+	FROM [%s]
+	WHERE lease_holder != 0`, q), [][]string{{"true"}})
 		})
 	}
 }

